pkg/i2gw/provider.go

```diff
@@ -99,8 +101,90 @@ type GatewayResources struct {
 	UDPRoutes  map[types.NamespacedName]gatewayv1alpha2.UDPRoute
 
 	ReferenceGrants map[types.NamespacedName]gatewayv1beta1.ReferenceGrant
+
+	GatewayExtensions map[types.NamespacedName]unstructured.Unstructured
+}
+
+// IR contains all Gateway-API objects, and Intermediate Representation(IR) to
+// hold specicifications on Gateway Objects for supporting Ingress extension
+// features. An feature on IR can be mapped to changes on Gateway-API objects,
+// or provider-specific Gateway extensions.
+type IR struct {
+	Gateways   map[types.NamespacedName]GatewayContext
+	HTTPRoutes map[types.NamespacedName]HTTPRouteContext
+	Services   map[types.NamespacedName]ServiceIR
+
+	GatewayClasses map[types.NamespacedName]gatewayv1.GatewayClass
+	TLSRoutes      map[types.NamespacedName]gatewayv1alpha2.TLSRoute
+	TCPRoutes      map[types.NamespacedName]gatewayv1alpha2.TCPRoute
+	UDPRoutes      map[types.NamespacedName]gatewayv1alpha2.UDPRoute
+
+	ReferenceGrants map[types.NamespacedName]gatewayv1beta1.ReferenceGrant
 }
 
+// GatewayContext contains the Gateway-API Gateway object, along with a
+// dedicated field for each provider to specify their extension features on
+// Gateway.
+type GatewayContext struct {
+	gatewayv1.Gateway
+
+	ApisixGatewayIR
```

Comment:
we probably need to not embed the extension fields and instead use
```
Apisix *ApisixGatewayIR
```

because 
* they are optional 
* the same fields may exist in both extensions

Reply:
Sounds good, updated